core/services/registrysyncer/syncer.go

```diff
@@ -27,19 +28,38 @@ type Syncer interface {
 	AddLauncher(h ...Launcher)
 }
 
+type ContractReaderFactory interface {
+	NewContractReader(context.Context, []byte) (types.ContractReader, error)
+}
+
+type RegistrySyncer interface {
+	Sync(ctx context.Context, isInitialSync bool) error
+	AddLauncher(launchers ...Launcher)
+	Start(ctx context.Context) error
+	Close() error
+	Ready() error
+	HealthReport() map[string]error
+	Name() string
+}
+
 type registrySyncer struct {
 	services.StateMachine
 	stopCh          services.StopChan
 	launchers       []Launcher
 	reader          types.ContractReader
-	initReader      func(ctx context.Context, lggr logger.Logger, relayer contractReaderFactory, registryAddress string) (types.ContractReader, error)
-	relayer         contractReaderFactory
+	initReader      func(ctx context.Context, lggr logger.Logger, relayer ContractReaderFactory, registryAddress string) (types.ContractReader, error)
+	relayer         ContractReaderFactory
 	registryAddress string
 	getPeerID       func() (p2ptypes.PeerID, error)
 
-	wg   sync.WaitGroup
-	lggr logger.Logger
-	mu   sync.RWMutex
+	orm ORM
+
+	updateChan chan *LocalRegistry
+
+	wg       sync.WaitGroup
+	lggr     logger.Logger
+	mu       sync.RWMutex
+	readerMu sync.RWMutex
```

Comment:
Which fields is this mutex guarding? Can't we just use "mu" but with a read-lock?

Reply:
It guards the `reader` field that we populate on `Sync` and read on `localRegistry`. We cannot use it, I tried it and it didn't fix the race. Upgrading the mutex on `Sync` from read to write hangs the test also, that's why I went with a separate lock for the reader setup.